### PR TITLE
Initial support for GPU channels

### PR DESCRIPTION
**Pull request by gdkchan**
This adds some basic support for GPU channels. Each channel has *separate state*. Due to the fact that everything was using a single channel before, games that made use of multiple Queues (each Queue is a channel) would end with corrupt state. The most common situation that causes this is Queue A setting up its state, then a new queue (Queue B) being created, and then setting up its own state, and corrupting Queues A state in the process.

This fixes the problem by adding a new `GpuChannel` class, and moving almost all of the state that shouldn't be shared there. A few things such as draw state were not moved yet, but those shouldn't cause issues as this state is reset often. GPU address spaces are also still shared, but this is a separate issue, and queues can share an address space (address spaces are created per Device rather than per Queue).

Things that still need to be done:
- ~~Separate host state. OpenGL doesn't allow this, at least not easily. The backend could expose a `CreatePipeline` method, and allow the creation of separate pipelines. Supporting this on OpenGL would require keeping track of each pipeline state and setting it on switch as needed, or using multiple contexts. Vulkan allows the creation of multiple queues, where each queue has separate state, but queues have a limit that varies per vendor.~~ This has been solved by forcing the host state to be updated on every channel switch.
- Separate remaining state on frontend (mostly draw related state).
- Separate GPU address spaces (this is likely the most complicated part). Likely requires the memory tracking to work over physical memory instead, otherwise the the caches are dependant on the guest process address space, which means that those resources can't be shared between processes.

This fixes a number of games that previously had `InvalidMemoryRegionException`s caused by addresses from already closed channels being used, due to all queues sharing the same channel.

Fixes https://github.com/Ryujinx/Ryujinx-Games-List/issues/2901
Now goes ingame.
image
image
image
(Note: Still has an issue when a save is loaded where the scene is mostly black. Needs further investigation.)

Fixes https://github.com/Ryujinx/Ryujinx-Games-List/issues/1965
Now goes ingame.
image
image
image
image

Fixes https://github.com/Ryujinx/Ryujinx-Games-List/issues/3415
Now goes ingame.
image
image
image
image
image
image

Likely also fixes (needs re-testing, I don't have those games):
https://github.com/Ryujinx/Ryujinx-Games-List/issues/82
https://github.com/Ryujinx/Ryujinx-Games-List/issues/745
https://github.com/Ryujinx/Ryujinx-Games-List/issues/924
https://github.com/Ryujinx/Ryujinx-Games-List/issues/3077
Might also fix Nekopara vol. 1/2/3 save issue?

### CODE REVIEW

File: Ryujinx.Graphics.Gpu/Memory/BufferCache.cs [hunk 1]
```diff
@@ -0,0 +1,391 @@
...
+using Ryujinx.Memory.Range;
+using System;
+using System.Collections.Generic;
+
+namespace Ryujinx.Graphics.Gpu.Memory
+{
+    /// <summary>
+    /// Buffer cache.
+    /// </summary>
+    class BufferCache
```
Comment (Thog) [inline Ryujinx.Graphics.Gpu/Memory/BufferCache.cs:13]: Missing ``IDisposable``

File: Ryujinx.Graphics.Gpu/Image/TextureManager.cs [hunk 2]
```diff
@@ -1,36 +1,13 @@
...
-using Ryujinx.Graphics.Texture;
-using Ryujinx.Memory.Range;
-using System;
 
 namespace Ryujinx.Graphics.Gpu.Image
 {
     /// <summary>
     /// Texture manager.
     /// </summary>
-    class TextureManager : IDisposable
```
Comment (Thog) [inline Ryujinx.Graphics.Gpu/Image/TextureManager.cs:15]: TextureManager still implement Dispose, this shouldn't be removed.

**Review by Thog — CHANGES_REQUESTED** [review]
Some tiny issues I found on my first review pass 👍 

So far I didn't found any regressions on the games I tested.

**Review by Thog — APPROVED** [review]

**Comment by zombie167** [comment]
It should be certain that Dragon Ball Xenoverse 2 can enter the game.
test

Final Fantasy X/X-2 HD Remaster not being able to enter the game should be a regression, because the earlier version 1.0.5796 is ok, this PR has not been fixed yet, I believe it is caused by some reason
https://youtu.be/j0lHliNHRhc

**Comment by EmulationFanatic** [comment]
> It should be certain that Dragon Ball Xenoverse 2 can enter the game.
> test
> 
> Final Fantasy X/X-2 HD Remaster not being able to enter the game should be a regression, because the earlier version 1.0.5796 is ok, this PR has not been fixed yet, I believe it is caused by some reason
> https://youtu.be/j0lHliNHRhc

Final Fantasy X/X-2 HD Remaster is not a regression on this PR. It does not go in-game on master and has not for some time, as noted in the issue:
https://github.com/Ryujinx/Ryujinx-Games-List/issues/876

**Comment by zombie167** [comment]
> > It should be certain that Dragon Ball Xenoverse 2 can enter the game.
> > test
> > Final Fantasy X/X-2 HD Remaster not being able to enter the game should be a regression, because the earlier version 1.0.5796 is ok, this PR has not been fixed yet, I believe it is caused by some reason
> > https://youtu.be/j0lHliNHRhc
> 
> Final Fantasy X/X-2 HD Remaster is not a regression on this PR. It does not go in-game on master and has not for some time, as noted in the issue:
> [Ryujinx/Ryujinx-Games-List#876](https://github.com/Ryujinx/Ryujinx-Games-List/issues/876)

Sorry for misunderstanding,I am not referring to the emulation regression caused by PR

**Comment by gdkchan** [comment]
It has come to my attention that `COTTOn Reboot!` was able to progress further with those changes, but didn't render anything, so I decided to take a look. It turns out that the game was broken because the host state was not updated, and one channel was using stale state from another one. I fixed it by marking all state as modified on channel switch, and also forcing rebind of all buffers, textures and render targets. This also solves one of the bullets of the "Things that still need to be done". The PR description has been updated and now includes test results of this game.

Anyone that already reviewed may wish to take a look at the latest commit, and maybe do a retest.

**Review by Thog — APPROVED** [review]

File: Ryujinx.Graphics.Gpu/Engine/GPFifo/GPFifoProcessor.cs [hunk 3]
```diff
@@ -186,5 +187,17 @@ public void SetShadowRamControl(ShadowRamControl control)
                 _subChannels[i].ShadowRamControl = control;
             }
         }
+
+        /// <summary>
+        /// Forces a full host state update by marking all state as modfied,
```
Comment (riperiperi) [inline Ryujinx.Graphics.Gpu/Engine/GPFifo/GPFifoProcessor.cs:192]: `modified`

**Review by riperiperi — APPROVED** [review]
Looking great, very nice work. Just one small nit.

**Review by riperiperi — APPROVED** [review]

**Pull request merged**